### ISSUE TITLE
IExecutorService#submitToAllMembers submits to already gracefully shutdown member

### ISSUE THREAD

**Issue by iboar** [hazelcast/hazelcast#12583]
Hello,

I have prepared a small demo and description which show what is the problem.
At first hazelcast configuration:

```
    private static Config configuration(boolean liteMember) {
        return new Config()
                .setProperty("hazelcast.logging.type", "slf4j")
                .setLiteMember(liteMember)
                .setNetworkConfig(networkConf())
                .setSerializationConfig(kryoSerializationConfiguration())
                .addMapConfig(new MapConfig("DummyMap")
                        .setInMemoryFormat(InMemoryFormat.BINARY)
                        .setReadBackupData(true)
                        .setBackupCount(1)
                );
    }

    private static NetworkConfig networkConf() {
        String resolvedClusterHost;
        try {
            resolvedClusterHost = getLocalHost().getHostAddress();
        } catch (UnknownHostException e) {
            throw new RuntimeException(e);
        }

        return new NetworkConfig()
                .setJoin(new JoinConfig()
                        .setMulticastConfig(new MulticastConfig()
                                .setMulticastGroup("224.2.2.3")
                                .setMulticastPort(7777)
                                .setMulticastTimeToLive(1)
                                .setEnabled(true)
                        )
                )
                .setPublicAddress(resolvedClusterHost)
                .setInterfaces(new InterfacesConfig()
                        .addInterface(resolvedClusterHost)
                        .setEnabled(true)
                );
    }

    private static SerializationConfig kryoSerializationConfiguration() {
        return new SerializationConfig()
                .setGlobalSerializerConfig(new GlobalSerializerConfig()
                        .setClassName(Serializer.class.getName())
                        .setOverrideJavaSerialization(true)
                );
    }
```

**What I do in my demo:**
- run Hazelcast lite member (A-node)
- wait until A-node is fully started
- run Hazelcast member (B-node which is not lite)
- wait until B-node finish it's work.

**What A-node does:**
- sleeps 20 seconds
- shutdowns itself gracefully using `HazelcastInstance#shutdown`
- logs "node is down - exiting"
- exits

**What B-node does:**
- reads in loop A-node logfile until "node is down - exiting" appears
- sleeps 2 seconds
- submits simple Callable to all members using `IExecutorService#submitToAllMembers`
- waits for results
- sleeps 20 seconds and exits

Main and help methods

```
public static void main(String[] args) throws ParseException, InterruptedException, IOException {
        final boolean liteMember = liteMemberRequested(args);
        HazelcastInstance hazelcast = createHazelcastInstance(liteMember);

        if (liteMember) {
            log.info("Lite member graceful shutdown in 20 seconds");
            sleepSeconds(20);
            hazelcast.shutdown();
            log.info("node is down - exiting");
        } else {
            waitUntilLiteMemberIsDown();
            sleepSeconds(2);
            submitTaskToAllMembers(hazelcast);
            sleepSeconds(20);
        }
        System.exit(0);
    }

    private static boolean liteMemberRequested(String[] args) throws ParseException {
        Options options = new Options();
        options.addOption("lite", false, "Starts node as Lite Member");
        CommandLine cmd = new DefaultParser().parse(options, args);
        return cmd.hasOption("lite");
    }

    @SuppressWarnings("StatementWithEmptyBody")
    private static void waitUntilLiteMemberIsDown() throws IOException {
        String memberIsDownMarker = "node is down - exiting";
        File liteMemberLogFile = new File("D:\\IdeaProjects\\Hazelcast-poc\\A.log");

        while (!readFileToString(liteMemberLogFile).contains(memberIsDownMarker)) ;
        log.info("detected that lite member is down");
    }

    private static void submitTaskToAllMembers(HazelcastInstance hazelcast) throws IOException {
        log.info("submitting task to all members");
        IExecutorService executorService = hazelcast.getExecutorService("executor-service");
        Collection<Future<Long>> futures = executorService.submitToAllMembers(new Task()).values();
        waitForResults(futures);
    }

    private static void waitForResults(Collection<Future<Long>> futures) {
        for (Future<Long> f : futures) {
            try {
                f.get();
            } catch (Exception e) {
                log.error("error while waiting for response", e);
                System.exit(2);
            }
        }
    }

    private static void sleepSeconds(long seconds) {
        try {
            sleep(seconds * 1000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    private static class Task implements Callable<Long>, Serializable, HazelcastInstanceAware {
        private final static Logger log = LoggerFactory.getLogger(Task.class);
        private transient HazelcastInstance hazelcast;

        @Override
        public void setHazelcastInstance(HazelcastInstance hazelcastInstance) {
            this.hazelcast = hazelcastInstance;
        }

        @Override
        public Long call() throws Exception {
            log.info("Running task");
            return hazelcast.getMap("DummyMap").getLocalMapStats().getHeapCost();
        }
    }
```

What happens:
- B-node detects that A-node exited and after 2 seconds send task to members, should send it only to itself but task is also sent to already shutdown member so it fails.

Please see logs below:
A-node log:
```
16:02:39,322  INFO ClusterService:65 - [192.168.99.1]:5701 [dev] [3.9.3] 

Members {size:1, ver:1} [
	Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad this lite
]

16:02:39,359  INFO LifecycleService:65 - [192.168.99.1]:5701 [dev] [3.9.3] [192.168.99.1]:5701 is STARTED
16:02:39,361  INFO Test:41 - Started Hazelcast Node
16:02:39,364  INFO Test:95 - Lite member graceful shutdown in 20 seconds
16:02:44,160  INFO TcpIpAcceptor:65 - [192.168.99.1]:5701 [dev] [3.9.3] Accepting socket connection from /192.168.99.1:51809
16:02:44,162  INFO TcpIpConnectionManager:65 - [192.168.99.1]:5701 [dev] [3.9.3] Established socket connection between /192.168.99.1:5701 and /192.168.99.1:51809
16:02:50,165  INFO ClusterService:65 - [192.168.99.1]:5701 [dev] [3.9.3] 

Members {size:2, ver:2} [
	Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad this lite
	Member [192.168.99.1]:5702 - 14958340-8ca5-4f1e-a407-2dd1f12afeff
]

16:02:59,364  INFO LifecycleService:65 - [192.168.99.1]:5701 [dev] [3.9.3] [192.168.99.1]:5701 is SHUTTING_DOWN
16:02:59,366  INFO Node:65 - [192.168.99.1]:5701 [dev] [3.9.3] Shutting down multicast service...
16:02:59,367  INFO Node:65 - [192.168.99.1]:5701 [dev] [3.9.3] Shutting down connection manager...
16:02:59,368  INFO TcpIpConnection:65 - [192.168.99.1]:5701 [dev] [3.9.3] Connection[id=1, /192.168.99.1:5701->/192.168.99.1:51809, endpoint=[192.168.99.1]:5702, alive=false, type=MEMBER] closed. Reason: TcpIpConnectionManager is stopping
16:02:59,371  INFO Node:65 - [192.168.99.1]:5701 [dev] [3.9.3] Shutting down node engine...
16:02:59,376  INFO NodeExtension:65 - [192.168.99.1]:5701 [dev] [3.9.3] Destroying node NodeExtension.
16:02:59,376  INFO Node:65 - [192.168.99.1]:5701 [dev] [3.9.3] Hazelcast Shutdown is completed in 10 ms.
16:02:59,446  INFO LifecycleService:65 - [192.168.99.1]:5701 [dev] [3.9.3] [192.168.99.1]:5701 is SHUTDOWN
16:02:59,446  INFO Test:98 - node is down - exiting
```

B-node log:
```
16:02:50,181  INFO ClusterService:65 - [192.168.99.1]:5702 [dev] [3.9.3] 

Members {size:2, ver:2} [
	Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad lite
	Member [192.168.99.1]:5702 - 14958340-8ca5-4f1e-a407-2dd1f12afeff this
]

16:02:52,217  INFO LifecycleService:65 - [192.168.99.1]:5702 [dev] [3.9.3] [192.168.99.1]:5702 is STARTED
16:02:52,217  INFO Test:41 - Started Hazelcast Node
16:02:52,479  INFO HealthMonitor:65 - [192.168.99.1]:5702 [dev] [3.9.3] processors=4, physical.memory.total=11,9G, physical.memory.free=5,6G, swap.space.total=23,9G, swap.space.free=12,9G, heap.memory.used=125,5M, heap.memory.free=354,0M, heap.memory.total=479,5M, heap.memory.max=479,5M, heap.memory.used/total=26,17%, heap.memory.used/max=26,17%, minor.gc.count=0, minor.gc.time=0ms, major.gc.count=0, major.gc.time=0ms, load.process=75,55%, load.system=100,00%, load.systemAverage=n/a thread.count=38, thread.peakCount=38, cluster.timeDiff=-1, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=7, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0,00%, operations.pending.invocations.count=0, proxy.count=0, clientEndpoint.count=0, connection.active.count=1, client.connection.count=0, connection.count=1
16:02:59,370  INFO TcpIpConnection:65 - [192.168.99.1]:5702 [dev] [3.9.3] Connection[id=1, /192.168.99.1:51809->/192.168.99.1:5701, endpoint=[192.168.99.1]:5701, alive=false, type=MEMBER] closed. Reason: Connection closed by the other side
16:02:59,372  INFO TcpIpConnector:65 - [192.168.99.1]:5702 [dev] [3.9.3] Connecting to /192.168.99.1:5701, timeout: 0, bind-any: true
16:02:59,447  INFO Test:121 - detected that lite member is down
16:03:00,374  INFO TcpIpConnector:65 - [192.168.99.1]:5702 [dev] [3.9.3] Could not connect to: /192.168.99.1:5701. Reason: SocketException[Connection refused: connect to address /192.168.99.1:5701]
16:03:00,477  INFO TcpIpConnector:65 - [192.168.99.1]:5702 [dev] [3.9.3] Connecting to /192.168.99.1:5701, timeout: 0, bind-any: true
16:03:01,448  INFO Test:125 - submitting task to all members
16:03:01,513  INFO TcpIpConnector:65 - [192.168.99.1]:5702 [dev] [3.9.3] Could not connect to: /192.168.99.1:5701. Reason: SocketException[Connection refused: connect to address /192.168.99.1:5701]
16:03:01,519  INFO TcpIpConnector:65 - [192.168.99.1]:5702 [dev] [3.9.3] Connecting to /192.168.99.1:5701, timeout: 0, bind-any: true
16:03:01,522  INFO Test$Task:161 - Running task
16:03:02,521  INFO TcpIpConnector:65 - [192.168.99.1]:5702 [dev] [3.9.3] Could not connect to: /192.168.99.1:5701. Reason: SocketException[Connection refused: connect to address /192.168.99.1:5701]
16:03:02,521  WARN TcpIpConnectionErrorHandler:67 - [192.168.99.1]:5702 [dev] [3.9.3] Removing connection to endpoint [192.168.99.1]:5701 Cause => java.net.SocketException {Connection refused: connect to address /192.168.99.1:5701}, Error-Count: 5
16:03:02,522  WARN MembershipManager:67 - [192.168.99.1]:5702 [dev] [3.9.3] Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad lite is suspected to be dead for reason: No connection
16:03:02,523  INFO MembershipManager:65 - [192.168.99.1]:5702 [dev] [3.9.3] Starting mastership claim process...
16:03:02,523  INFO MembershipManager:65 - [192.168.99.1]:5702 [dev] [3.9.3] Local MembersView{version=2, members=[MemberInfo{address=[192.168.99.1]:5701, uuid=9ac3a6e4-694f-43bb-9307-3b81268a22ad, liteMember=true}, MemberInfo{address=[192.168.99.1]:5702, uuid=14958340-8ca5-4f1e-a407-2dd1f12afeff, liteMember=false}]} with suspected members: [[192.168.99.1]:5701] and initial addresses to ask: []
16:03:02,528  INFO ClusterService:65 - [192.168.99.1]:5702 [dev] [3.9.3] 

Members {size:1, ver:3} [
	Member [192.168.99.1]:5702 - 14958340-8ca5-4f1e-a407-2dd1f12afeff this
]

16:03:02,529  INFO MembershipManager:65 - [192.168.99.1]:5702 [dev] [3.9.3] Mastership is claimed with: MembersView{version=3, members=[MemberInfo{address=[192.168.99.1]:5702, uuid=14958340-8ca5-4f1e-a407-2dd1f12afeff, liteMember=false}]}
16:03:02,529  INFO TransactionManagerService:65 - [192.168.99.1]:5702 [dev] [3.9.3] Committing/rolling-back alive transactions of Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad lite, UUID: 9ac3a6e4-694f-43bb-9307-3b81268a22ad
16:03:02,530 ERROR Test:136 - error while waiting for response
com.hazelcast.core.MemberLeftException: Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad lite has left cluster!
```

- A-node is already shutdown at 16:02:59,446
- after 1 millisecond B-node knows about it
- after 2 seconds at 16:03:01,448 B-node sends task to all members
- at 16:03:02,529 cluster really knows that member has left. After more than 3 seconds on same machine.

I would expect when member is shutdown gracefully cluster knows that any tasks submitted to all members should not be sent to this specific member.

Could you give me any suggestion how to avoid situations like that?

Regards

**Comment by mdogan** [comment]
I don't think there's an issue here. All members in the cluster are not guaranteed to observe leave/removal of another member _instantly_.  They will eventually realize that a member is left at some point. 

According to your logs, your log reader mechanism observes exit of A-node at `16:02:59,447`.
```
16:02:59,447  INFO Test:121 - detected that lite member is down
```

Then B-node submitted task to all members at `16:03:01,448`
```
16:03:01,448  INFO Test:125 - submitting task to all members
```

Then B-node realized leave of A-node at `16:03:02,522`
```
16:03:02,521  WARN TcpIpConnectionErrorHandler:67 - [192.168.99.1]:5702 [dev] [3.9.3] Removing connection to endpoint [192.168.99.1]:5701 Cause => java.net.SocketException {Connection refused: connect to address /192.168.99.1:5701}, Error-Count: 5
16:03:02,522  WARN MembershipManager:67 - [192.168.99.1]:5702 [dev] [3.9.3] Member [192.168.99.1]:5701 - 9ac3a6e4-694f-43bb-9307-3b81268a22ad lite is suspected to be dead for reason: No connection
```

**Comment by mmedenjak** [comment]
Hi @gdela !

We may consider this improvement in the future but for now node B is not responsible for maintaining the membership list. If it were, there might be some other issues we'd run into and we need to consider this change in depth. Even if node B would send and information to other members, there's still a race. Node B might have sent the "don't call me, I'll be down" message just after the task has been scheduled or the messages might have been reordered by the network.

For now, the membership list is maintained by the master node and as @mdogan mentioned, other nodes eventually see the correct membership list. As such, there is a race between your thread detecting that the node has been terminated and node B which determines it through the membership list dissemination.

Bottom line - we might consider some improvements on this but for now we will keep this behaviour.
